File: backend/utils/delivery_day_detect.py
"""Detección de día de entrega desde observaciones OC (pre-despacho)."""

from __future__ import annotations

import re
import unicodedata

DAY_TOKENS: tuple[str, ...] = (
    "lunes",
    "martes",
    "miercoles",
    "jueves",
    "viernes",
    "sabado",
    "domingo",
)

DAY_TOKEN_SET = frozenset(DAY_TOKENS)

DAY_LABEL: dict[str, str] = {
    "lunes": "Lunes",
    "martes": "Martes",
    "miercoles": "Miércoles",
    "jueves": "Jueves",
    "viernes": "Viernes",
    "sabado": "Sábado",
    "domingo": "Domingo",
}

_DAY_RE = re.compile(
    r"\b(" + "|".join(DAY_TOKENS) + r")\b",
    re.IGNORECASE,
)

_DELIVERY_CTX_RE = re.compile(
    r"\b(entrega|retiro|reparto|despacho)\b",
    re.IGNORECASE,
)

# Claves sin tildes → etiqueta canónica comuna
MUNICIPALITY_CANONICAL: dict[str, str] = {
    "ancud": "Ancud",
    "quellon": "Quellón",
    "castro": "Castro",
    "chonchi": "Chonchi",
    "dalcahue": "Dalcahue",
    "achao": "Achao",
    "queilen": "Queilén",
    "quemchi": "Quemchi",
    "puqueldon": "Puqueldón",
    "melinka": "Melinka",
}


def strip_accents(text: str) -> str:
    nk = unicodedata.normalize("NFD", text)
    return "".join(c for c in nk if unicodedata.category(c) != "Mn")


def normalize_day_token(raw: str | None) -> str | None:
    if raw is None or not str(raw).strip():
        return None
    s = strip_accents(str(raw).strip().lower())
    s = "".join(c for c in s if c.isalpha())
    return s if s in DAY_TOKEN_SET else None


def normalize_municipality_name(raw: str | None) -> str:
    t = (raw or "").strip()
    if not t:
        return "Sin comuna"
    key = strip_accents(t).lower()
    if key in MUNICIPALITY_CANONICAL:
        return MUNICIPALITY_CANONICAL[key]
    return t[0].upper() + t[1:] if len(t) > 1 else t.upper()


def delivery_day_label(token: str | None) -> str:
    if not token:
        return "Sin día"
    return DAY_LABEL.get(token, token.capitalize())


def detect_delivery_day_from_observation(text: str | None) -> str | None:
    """
    Extrae un único día desde texto de observación.
    Prioriza menciones cercanas a entrega/retiro; si hay varias, la última en el texto.
    """
    if text is None or not str(text).strip():
        return None
    norm = strip_accents(str(text).strip().lower())
    if not norm:
        return None

    matches = list(_DAY_RE.finditer(norm))
    if not matches:
        return None
    if len(matches) == 1:
        return normalize_day_token(matches[0].group(1))

    # Varias menciones: preferir la última con contexto de entrega; si ninguna, la última del texto.
    best: str | None = None
    best_score = -1.0
    last_token: str | None = None
    for m in matches:
        token = normalize_day_token(m.group(1))
        if not token:
            continue
        last_token = token
        start, end = m.span()
        window = norm[max(0, start - 48) : min(len(norm), end + 48)]
        score = float(start) / max(len(norm), 1)
        if _DELIVERY_CTX_RE.search(window):
            score += 10.0
        if score >= best_score:
            best_score = score
            best = token
    return best or last_token


def resolve_delivery_day(
    observaciones: str | None,
    comments: str | None = None,
    dia_atencion: str | None = None,
) -> tuple[str | None, str]:
    """
    Resuelve día de entrega desde fuentes Bsale actuales.
    1) Día en atributo OBSERVACIONES (más reciente)
    2) Día en comentarios del documento
    3) Texto combinado (obs + comentarios) — última mención con contexto
    4) Ruta / cliente (dia_atencion)
    """
    obs_text = (observaciones or "").strip()
    comments_text = (comments or "").strip()

    obs_day = detect_delivery_day_from_observation(obs_text)
    if obs_day:
        return obs_day, "observacion"

    comments_day = detect_delivery_day_from_observation(comments_text)
    if comments_day:
        return comments_day, "comentario"

    combined = "\n".join(x for x in (obs_text, comments_text) if x)
    combined_day = detect_delivery_day_from_observation(combined)
    if combined_day:
        return combined_day, "observacion"

    route = normalize_day_token(dia_atencion)
    if route:
        return route, "ruta"
    return None, "sin_dia"


def delivery_day_matches_filter(
    detected: str | None,
    filter_tokens: list[str],
) -> bool:
    if not filter_tokens:
        return True
    if not detected:
        return False
    return detected in filter_tokens


# SQL: última mención de día con límite de palabra (PostgreSQL \m \M).
_DETECT_DAY_SQL_INNER = r"""(regexp_match(
    translate(lower(COALESCE({expr}, '')), 'áéíóúü', 'aeiouu'),
    '.*\m(lunes|martes|miercoles|jueves|viernes|sabado|domingo)\M'
))[1]"""


def sql_detect_delivery_day(expr: str) -> str:
    return _DETECT_DAY_SQL_INNER.format(expr=expr)


def sql_resolve_delivery_day(
    observaciones_expr: str,
    comments_expr: str,
    dia_atencion_expr: str,
) -> str:
    """Misma precedencia que ``resolve_delivery_day`` (Python).

    1) día en observaciones
    2) día en comentarios (aunque observaciones tenga texto sin día)
    3) día de ruta / cliente
    """
    obs_det = sql_detect_delivery_day(observaciones_expr)
    comments_det = sql_detect_delivery_day(comments_expr)
    route_det = sql_detect_delivery_day(dia_atencion_expr)
    return f"""COALESCE(
    {obs_det},
    {comments_det},
    {route_det}
)"""

File: frontend/lib/delivery-day-detect.ts
/**
 * Día de entrega detectado desde observaciones OC (pre-despacho).
 * Espejo de backend/utils/delivery_day_detect.py
 */

const DAY_TOKENS = [
  "lunes",
  "martes",
  "miercoles",
  "jueves",
  "viernes",
  "sabado",
  "domingo",
] as const

export type DeliveryDayToken = (typeof DAY_TOKENS)[number]

const DAY_LABEL: Record<DeliveryDayToken, string> = {
  lunes: "Lunes",
  martes: "Martes",
  miercoles: "Miércoles",
  jueves: "Jueves",
  viernes: "Viernes",
  sabado: "Sábado",
  domingo: "Domingo",
}

const DAY_RE = new RegExp(`\\b(${DAY_TOKENS.join("|")})\\b`, "gi")
const DELIVERY_CTX_RE = /\b(entrega|retiro|reparto|despacho)\b/i

function stripAccents(s: string): string {
  return s
    .toLowerCase()
    .normalize("NFD")
    .replace(/\p{M}/gu, "")
}

function normalizeDayToken(raw: string | null | undefined): DeliveryDayToken | null {
  if (!raw?.trim()) return null
  const s = stripAccents(raw).replace(/[^a-z]/g, "")
  return (DAY_TOKENS as readonly string[]).includes(s) ? (s as DeliveryDayToken) : null
}

export function detectDeliveryDayFromObservation(
  text: string | null | undefined,
): DeliveryDayToken | null {
  if (!text?.trim()) return null
  const norm = stripAccents(text)
  const matches = [...norm.matchAll(DAY_RE)]
  if (matches.length === 0) return null
  if (matches.length === 1) return normalizeDayToken(matches[0][1])

  let best: DeliveryDayToken | null = null
  let bestScore = -1
  for (const m of matches) {
    const token = normalizeDayToken(m[1])
    if (!token) continue
    const idx = m.index ?? 0
    const window = norm.slice(Math.max(0, idx - 48), idx + m[0].length + 48)
    let score = idx / Math.max(norm.length, 1)
    if (DELIVERY_CTX_RE.test(window)) score += 10
    if (score >= bestScore) {
      bestScore = score
      best = token
    }
  }
  return best
}

export function formatDeliveryDayLabel(
  token: string | null | undefined,
): string {
  if (!token?.trim()) return "Sin día"
  const k = normalizeDayToken(token)
  if (!k) return "Sin día"
  return DAY_LABEL[k]
}

const DAY_BADGE_CLASS: Record<DeliveryDayToken, string> = {
  lunes: "border-slate-300 bg-slate-50 text-slate-800 dark:bg-slate-900/40",
  martes: "border-slate-300 bg-slate-50 text-slate-800 dark:bg-slate-900/40",
  miercoles: "border-blue-200 bg-blue-50 text-blue-900 dark:bg-blue-950/40",
  jueves: "border-indigo-200 bg-indigo-50 text-indigo-900 dark:bg-indigo-950/40",
  viernes: "border-violet-200 bg-violet-50 text-violet-900 dark:bg-violet-950/40",
  sabado: "border-amber-300 bg-amber-50 text-amber-950 dark:bg-amber-950/40",
  domingo: "border-rose-200 bg-rose-50 text-rose-900 dark:bg-rose-950/40",
}

export function deliveryDayBadgeClass(token: string | null | undefined): string {
  const k = normalizeDayToken(token)
  if (!k) return "border-border bg-muted/40 text-muted-foreground"
  return DAY_BADGE_CLASS[k]
}

/** Etiqueta visual para tabla pre-despacho (incluye Retiro si aplica). */
export function formatPreDespachoDeliveryDay(row: {
  observaciones?: string | null
  dia_entrega_detectado?: string | null
  dia_entrega_label?: string | null
}): string {
  const base =
    row.dia_entrega_label?.trim() ||
    formatDeliveryDayLabel(row.dia_entrega_detectado)
  const obs = (row.observaciones || "").toLowerCase()
  if (/\bretiro\b/.test(obs) && base !== "Sin día") {
    return `Retiro · ${base}`
  }
  if (/\bretiro\b/.test(obs) && base === "Sin día") {
    return "Retiro"
  }
  return base
}
